Win32 code refresh & audio buffer enhancements

sukibaby opened the pull request:
The old PR #255 has lasted through a large number of changes to the `beta` branch, as a result the history became very messy and could not be rebased or squashed. For this reason I have made a new branch from the finalized state of that PR so everything can live in one commit.

### Overview
The goal of this PR is to solve the problem of unpredictable glitches on Windows hosts. On Windows, many time critical threads, including all rendering threads, are run at normal priority, meaning the system views them as very unimportant. 

When going through the code to resolve these issues, a large amount of Windows 9x compatibility code was found, and replaced with modernized code compatible with Windows 7 and up.

Bad code in the sound mixing buffer would lead to runaway memory fragmentation, causing a significant amount of RAM to be wasted. The sound mixing buffer now pre-allocates a safe amount of memory which results in a significant decrease in RAM consumption (from ~1.4GB to ~450MB on my Windows 11 PC).

There is a separate sound buffer for the Windows audio driver, and this now pre-allocates a larger amount of memory to avoid frequent reallocations.

These changes are largely interdependent so they are all in one PR.

**GameLoop.cpp**
- Check if the OS is Windows, and if so ensure the rendering thread is set to the highest priority

**RageSound.cpp**
- Use `-1` instead of `RageSoundReader::ERROR` to prevent needing to undefine ERROR

**RageSoundMixBuffer.cpp / RageSoundMixBuffer.h**

- Set a buffer of 2MB instead of setting the buffer to nullptr and depend on it being grown by assets being loaded at the game launch
- Update some C style casts to C++ style casts
- Improve `Extend` feature to support handling a larger number of samples, and error handling in case of a memory allocation failure
    - I don't really like using std::exit in case the new buffer is nullptr, but I haven't decided what I would like to do here if it is nullptr

**RageSoundReader_Chain.cpp**

- Prevent code running needlessly on a null pointer

**RageSoundReader_ThreadedBuffer.cpp**

- BufferingThread(): Calculate unchanging values outside of the loop for better performance

**RageSoundReader_ThreadedBuffer.h**

- Include `windows.h` if the OS is Windows so that the sound buffer can be set to high priority

**arch/ArchHooks/ArchHooks_Win32.cpp**

- Set the main thread back to `ABOVE_NORMAL_PRIORITY_CLASS` to ensure the sound and rendering threads stay higher priority (reverts #228)

**arch/Sound/DSoundHelpers.cpp**

- Set the DirectSound buffer to 512KB instead of 64KB

**arch/Sound/RageSoundDriver_DSound_Software.cpp**

- Set the DirectSound decoding thread to `THREAD_PRIORITY_HIGHEST`

**arch/Sound/RageSoundDriver_WDMKS.cpp**

- Allow the buffer size to adapt to the number of channels and maximum bytes per sample
- Set one temporary buffer big enough to reuse multiple times instead of setting the buffer to ~2-4KB and reallocating it as needed
- Keep track of the starting position of `pBuf`
- Set the WDM decoding thread to `THREAD_PRIORITY_HIGHEST`

**arch/Threads/Threads_Win32.cpp**

- Rewrite `SetThreadName` to be GCC/GDB-friendly
- Update `StartThread` to use modern Windows features
- Update `MutexImpl_Win32::Lock` to use modern Windows features
- Replace all Win 9x compatibility code in `PortableSignalObjectAndWait`

sukibaby committed:
Win32 refresh & audio buffer fixes

GameLoop.cpp:
- Check if the OS is Windows, and if so ensure the rendering thread is set to the highest priority

RageSound.cpp
- Use -1 instead of RageSoundReader::ERROR to prevent needing to undefine ERROR

RageSoundMixBuffer.cpp / RageSoundMixBuffer.h
- Set a buffer of 2MB instead of setting the buffer to nullptr and depend on it being grown by assets being loaded at the game launch
- Update some C style casts to C++ style casts
- Improve Extend feature to support handling a larger number of samples, and error handling in case of a memory allocation failure

RageSoundReader_Chain.cpp
- Prevent code running needlessly on a null pointer

RageSoundReader_ThreadedBuffer.cpp
- BufferingThread(): Calculate unchanging values outside of the loop for better performance

RageSoundReader_ThreadedBuffer.h
- Include windows.h if the OS is Windows so that the sound buffer can be set to high priority

arch/ArchHooks/ArchHooks_Win32.cpp
- Set the main thread back to ABOVE_NORMAL_PRIORITY_CLASS to ensure the sound and rendering threads stay higher priority (reverts Change to HIGH_PRIORITY_CLASS on Windows #228)

arch/Sound/DSoundHelpers.cpp
- Set the DirectSound buffer to 512KB instead of 64KB

arch/Sound/RageSoundDriver_DSound_Software.cpp
- Set the DirectSound decoding thread to THREAD_PRIORITY_HIGHEST

arch/Sound/RageSoundDriver_WDMKS.cpp
- Allow the buffer size to adapt to the number of channels and maximum bytes per sample
- Set one temporary buffer big enough to reuse multiple times instead of setting the buffer to ~2-4KB and reallocating it as needed
- Keep track of the starting position of pBuf
- Set the WDM decoding thread to THREAD_PRIORITY_HIGHEST

arch/Threads/Threads_Win32.cpp
- Rewrite SetThreadName to be GCC/GDB-friendly
- Update StartThread to use modern Windows features
- Update MutexImpl_Win32::Lock to use modern Windows features
- Replace all Win 9x compatibility code in PortableSignalObjectAndWait

restructure priorities for windows with the understanding that the thread priorities should be structured as such:

Rendering thread, Audio mixing thread <- runs at Highest or Time Critical priority
Audio decoding thread, audio streaming thread <- Above Normal or Highest priority
Main game thread <- Normal or Above Normal priority

Putting stuff into constants

teejusb commented on src/RageSoundMixBuffer.cpp:
```diff
@@ -4,17 +4,30 @@
+
+	// Check if memory allocation was successful
+	if (m_pMixbuf == nullptr) {
+		ASSERT_M(false, "Sound mixing buffer being null should not be possible.");
```
I'd change the message to "Failed to allocate memory for the sound mixing buffer"

teejusb commented on src/RageSoundMixBuffer.cpp:
```diff
@@ -24,18 +37,26 @@ void RageSoundMixBuffer::SetWriteOffset( int iOffset )
 	{
-		m_pMixbuf = (float *) realloc( m_pMixbuf, sizeof(float) * realsize );
-		m_iBufSize = realsize;
+		float* m_pMixbufNew = static_cast<float*>(realloc(m_pMixbuf, sizeof(float) * realsize));
```
You can reorder these statements to remove local variable:

```cpp
m_pMixbuf = static_cast<float*>(realloc(m_pMixbuf, sizeof(float) * realsize));
if (m_pMixbuf == nullptr)
{
  ASSERT_M(false, "Failed to re-allocate memory for the sound mixing buffer.");
}
m_iBufSize = realsize;
```

teejusb commented on src/RageSoundReader_ThreadedBuffer.cpp:
```diff
@@ -191,53 +191,50 @@ bool RageSoundReader_ThreadedBuffer::SetProperty( const RString &sProperty, floa
-		int iRet = FillFrames( iFramesToFill );
+		int iRet = FillFrames(iFramesToFill);
 
-		// Release m_bFilling, and signal the event to wake anyone waiting for it.
```
Keep this comment

teejusb commented on src/RageSoundReader_ThreadedBuffer.cpp:
```diff
@@ -191,53 +191,50 @@ bool RageSoundReader_ThreadedBuffer::SetProperty( const RString &sProperty, floa
 		m_bFilling = false;
 		m_Event.Broadcast();
 
-		// On error or end of file, stop buffering the sound.
```
Keep this comment

teejusb commented on src/RageSoundReader_ThreadedBuffer.h:
```diff
@@ -70,7 +76,20 @@ class RageSoundReader_ThreadedBuffer: public RageSoundReader_Filter
-	static int StartBufferingThread( void *p ) { ((RageSoundReader_ThreadedBuffer *) p)->BufferingThread(); return 0; }
+	static int StartBufferingThread(void* p)
+	{
+#ifdef _WINDOWS
```
You have `#ifdef _WIN32` above. Which one is it?

teejusb reviewed (commented):
I haven't finished the review but can we change it so that we don't change whitespace in these PRs? It's a lot of noise and makes it hard to review as quickly as I'd like.

sukibaby commented:
> I haven't finished the review but can we change it so that we don't change whitespace in these PRs? It's a lot of noise and makes it hard to review as quickly as I'd like.

I will have the white space reverted on the next commit.

On that note, after this is merged, I was considering running the entire codebase thru Visual Studio's code cleanup feature, so that we don't have to be constantly putting back the old whitespace when using modern IDE's.

mjvotaw commented:
Yeah I think updating the formatting might be a good idea.
The only hesitation I have is that it would be obscuring a lot of git-blame information, but then again the "Line endings be normalized!" commit already did that for most of the codebase.

teejusb commented:
Formatting PR (that I admittedly do still need to finish) is here: https://github.com/itgmania/itgmania/pull/98 so I can handle it in the future.

sukibaby commented:
> Formatting PR (that I admittedly do still need to finish) is here: #98 so I can handle it in the future.

Np, let me know if you need a hand with it.

sukibaby commented:
> Yeah I think updating the formatting might be a good idea. The only hesitation I have is that it would be obscuring a lot of git-blame information, but then again the "Line endings be normalized!" commit already did that for most of the codebase.

I agree, that's my main hang up about it, but on the other hand manually going thru code after working in VS and putting back all the old whitespace is so miserable.

sukibaby commented:
I just noticed a `memset` and a `realloc` that I didn't change to have the `std::` prefix 🤦‍♀️  fixed now.

teejusb commented:
> > Yeah I think updating the formatting might be a good idea. The only hesitation I have is that it would be obscuring a lot of git-blame information, but then again the "Line endings be normalized!" commit already did that for most of the codebase.
> 
> I agree, that's my main hang up about it, but on the other hand manually going thru code after working in VS and putting back all the old whitespace is so miserable.

You can just disable format on save btw

sukibaby commented:
> > > Yeah I think updating the formatting might be a good idea. The only hesitation I have is that it would be obscuring a lot of git-blame information, but then again the "Line endings be normalized!" commit already did that for most of the codebase.
> > 
> > 
> > I agree, that's my main hang up about it, but on the other hand manually going thru code after working in VS and putting back all the old whitespace is so miserable.
> 
> You can just disable format on save btw

Uggghhhh I like using this feature when working on non-ITGm stuff 😔 I will consider it 

btw if you need anything from me or have any questions/concerns about this PR let me know or send me a message on discord 🙏

sukibaby commented:
Here is some additional context for these changes.

--------

Keep this comment from `src/arch/ArchHooks/ArchHooks_Win32.cpp` in mind:
https://github.com/itgmania/itgmania/blob/526c62daa68d525168a76d3d7e3c3e5b8fe12185/src/arch/ArchHooks/ArchHooks_Win32.cpp#L151-L155

In the codebase as-is, this is not at all how the priority of the threads are structured.

Windows requires the program to request a handle to the thread that the priority should be changed on, and then have the program pass along the desired priority. Thus any thread we wish to have at a priority besides "normal" must request a handle, which requires `windows.h` (`WIN32_LEAN_AND_MEAN` can be used to reduce the number of headers pulled in, however).

------

We can see that on the existing code base, the threads won't be boosted. `ITGmania plain.exe` is what I've named the release 0.8.0 version to.

![Screenshot 2024-06-05 212325](https://github.com/itgmania/itgmania/assets/163092272/435f815a-4fda-4415-a39a-79d96cccf3db)

We can see that after giving each relevant thread the ability to request the priority level it needs, the threads do in fact spawn at the intended priority.

![Screenshot 2024-06-05 212546](https://github.com/itgmania/itgmania/assets/163092272/b8838763-780a-4c00-9649-cecf6d4a4cb2)

These screenshots were taken mid-game. Here is a screenshot on the options menu, so there is no active music thread running.

![Screenshot 2024-06-05 212638](https://github.com/itgmania/itgmania/assets/163092272/a33cab74-0259-464b-ace9-9c5e49aa5876)


------

We can also see that in the existing code, there are provisions to do this, but not on the sound threads that are typically called. It's not clear why these functions can set the priority opt to raise them higher than the game thread, but out of all of them, only the WDMKS driver chooses a significantly priority than the game thread. Please note RageSound's ThreadedBuffer will default to using WaveOut, but when ThreadedBuffer is called it will be Highest priority, so it will elevate the priority of whatever driver it uses.

https://github.com/itgmania/itgmania/blob/903fa7eea399a0e4e27fe0b5bca32bfbdc21676b/src/arch/Sound/RageSoundDriver_WDMKS.cpp#L1157-L1162

https://github.com/itgmania/itgmania/blob/903fa7eea399a0e4e27fe0b5bca32bfbdc21676b/src/arch/Sound/RageSoundDriver_DSound_Software.cpp#L132-L136

https://github.com/itgmania/itgmania/blob/903fa7eea399a0e4e27fe0b5bca32bfbdc21676b/src/arch/Sound/RageSoundDriver_WaveOut.cpp#L50-L53

We can also see, with the new code, a considerable amount of time is saved on launching the game, even with all the delays caused by the debugger, and the new build handles the game window being moved around better while the debugger is active.

https://github.com/itgmania/itgmania/assets/163092272/1c3c158f-02ab-47e8-9b56-1147b0f4a787


https://github.com/itgmania/itgmania/assets/163092272/5822b1b6-410f-435a-ab28-fb7dde4649a6

sukibaby commented:
Added two improvements to this PR today.

1) I restructured the thread priorities in the new commit, otherwise everything else in this PR has remained the same on this latest update. This follows the theory that the thread priorities should be structured as such:


      1) Rendering thread, Audio mixing thread **<- should run at Highest or Time Critical priority**
      2) Audio decoding thread, audio streaming thread **<- should run at Highest priority**
      3) Main game thread **<- should run at Normal or Above Normal priority**
     4) Worker threads **<- should run at Below Normal or Normal priority** [worker threads were not changed in this PR]


2) Cleaned up the sound buffer initialization code, and make a constant of the size of a float as it's used in many calculations (https://github.com/sukibaby/itgmania/commit/0c042489b99022fba230cc20288da4dcc2e3208b)

teejusb commented:
Sorry for the delay but do you think it would be possible to break up this PR into a couple smaller ones. There seems be a couple of separate moving parts it seems to be addressing and it's a bit hard to follow. I keep trying to review it and then getting lost in the weeds >_>

sukibaby commented:
> Sorry for the delay but do you think it would be possible to break up this PR into a couple smaller ones. There seems be a couple of separate moving parts it seems to be addressing and it's a bit hard to follow. I keep trying to review it and then getting lost in the weeds >_>

No worries, i will do that! Please check back in a few hours.
